File: 1046-last-stone-weight/1046-last-stone-weight.cpp
class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {

        priority_queue<int> maxHeap;
        for (auto& stone : stones) {
            maxHeap.push(stone);
        }

        while (maxHeap.size() > 1) {
            int firstLargestStone = maxHeap.top();
            maxHeap.pop();

            int secondLargestStone = maxHeap.top();
            maxHeap.pop();

            int diff = firstLargestStone - secondLargestStone;

            if (diff) {
                maxHeap.push(diff);
            }
        }

        return maxHeap.size() == 1 ? maxHeap.top() : 0; 
        
    }
};

File: 1046-last-stone-weight/1046-last-stone-weight.py
class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:

        pq = []

        for stone in stones:
            heapq.heappush(pq, -stone)
        
        while len(pq) > 1:
            a = -heapq.heappop(pq)
            b = -heapq.heappop(pq)
            if a != b:
                heapq.heappush(pq, -(a - b))
        
        return -pq[0] if pq else 0
        